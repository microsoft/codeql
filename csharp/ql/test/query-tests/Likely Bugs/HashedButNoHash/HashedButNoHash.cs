using System.Collections;
using System.Collections.Generic;

public class Test
{
    public void M()
    {
        var h = new Hashtable();
        h.Add(this, null); // $ Alert
        h.Contains(this); // $ Alert
        h.ContainsKey(this); // $ Alert
        h[this] = null; // $ Alert
        h.Remove(this); // $ Alert

        var l = new List<Test>();
        l.Add(this); // Good

        var d = new Dictionary<Test, bool>();
        d.Add(this, false); // $ Alert
        d.ContainsKey(this); // $ Alert
        d[this] = false; // $ Alert
        d.Remove(this); // $ Alert
        d.TryAdd(this, false); // $ Alert
        d.TryGetValue(this, out bool _); // $ Alert

        var hs = new HashSet<Test>();
        hs.Add(this); // $ Alert
        hs.Contains(this); // $ Alert
        hs.Remove(this); // $ Alert
    }

    public override bool Equals(object other)
    {
        return false;
    }
}
