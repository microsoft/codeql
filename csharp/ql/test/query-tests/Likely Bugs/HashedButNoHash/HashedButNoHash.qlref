query: Likely Bugs/HashedButNoHash.ql
postprocess: utils/test/InlineExpectationsTestQuery.ql

