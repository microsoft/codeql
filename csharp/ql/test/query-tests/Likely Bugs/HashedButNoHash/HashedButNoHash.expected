| HashedButNoHash.cs:9:15:9:18 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:10:20:10:23 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:11:23:11:26 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:12:11:12:14 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:13:18:13:21 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:19:15:19:18 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:20:23:20:26 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:21:11:21:14 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:22:18:22:21 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:23:18:23:21 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:24:23:24:26 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:27:16:27:19 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:28:21:28:24 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
| HashedButNoHash.cs:29:19:29:22 | this access | This expression is hashed, but type 'Test' only defines Equals(...) not GetHashCode(). |
